File: recursion/2/sol.c
#include <stdio.h>
#include <stdlib.h>

// Definition for singly-linked list
struct ListNode {
    int val;
    struct ListNode *next;
};

// Function declaration
struct ListNode* addTwoNumbers(struct ListNode* l1, struct ListNode* l2);

// Helper function to create a new node
struct ListNode* createNode(int val) {
    struct ListNode* node = (struct ListNode*)malloc(sizeof(struct ListNode));
    node->val = val;
    node->next = NULL;
    return node;
}

// Helper function to create a linked list from an array
struct ListNode* createList(int* arr, int size) {
    if (size == 0) return NULL;
    
    struct ListNode dummy;
    struct ListNode* current = &dummy;
    
    for (int i = 0; i < size; i++) {
        current->next = createNode(arr[i]);
        current = current->next;
    }
    
    return dummy.next;
}

// Helper function to print a linked list
void printList(struct ListNode* head) {
    printf("[");
    while (head) {
        printf("%d", head->val);
        if (head->next) printf(",");
        head = head->next;
    }
    printf("]\n");
}

// Helper function to free a linked list
void freeList(struct ListNode* head) {
    struct ListNode* temp;
    while (head) {
        temp = head;
        head = head->next;
        free(temp);
    }
}

int main() {
    // Example 1
    printf("Example 1:\n");
    int arr1[] = {2, 4, 3};
    int arr2[] = {5, 6, 4};
    
    struct ListNode* l1 = createList(arr1, 3);
    struct ListNode* l2 = createList(arr2, 3);
    
    printf("Input: l1 = [2,4,3], l2 = [5,6,4]\n");
    struct ListNode* result1 = addTwoNumbers(l1, l2);
    printf("Output: ");
    printList(result1);
    printf("Expected: [7,0,8]\n\n");
    
    // Example 2
    printf("Example 2:\n");
    int arr3[] = {0};
    int arr4[] = {0};
    
    struct ListNode* l3 = createList(arr3, 1);
    struct ListNode* l4 = createList(arr4, 1);
    
    printf("Input: l1 = [0], l2 = [0]\n");
    struct ListNode* result2 = addTwoNumbers(l3, l4);
    printf("Output: ");
    printList(result2);
    printf("Expected: [0]\n\n");
    
    // Example 3
    printf("Example 3:\n");
    int arr5[] = {9, 9, 9, 9, 9, 9, 9};
    int arr6[] = {9, 9, 9, 9};
    
    struct ListNode* l5 = createList(arr5, 7);
    struct ListNode* l6 = createList(arr6, 4);
    
    printf("Input: l1 = [9,9,9,9,9,9,9], l2 = [9,9,9,9]\n");
    struct ListNode* result3 = addTwoNumbers(l5, l6);
    printf("Output: ");
    printList(result3);
    printf("Expected: [8,9,9,9,0,0,0,1]\n");
    
    // Free memory
    freeList(l1);
    freeList(l2);
    freeList(l3);
    freeList(l4);
    freeList(l5);
    freeList(l6);
    freeList(result1);
    freeList(result2);
    freeList(result3);
    
    return 0;
}

File: linkedlist/2/sol.py
from my import Solution, ListNode

def list_to_linked_list(lst):
    """Convert a Python list to a linked list"""
    if not lst:
        return None
    
    dummy = ListNode(0)
    current = dummy
    
    for val in lst:
        current.next = ListNode(val)
        current = current.next
    
    return dummy.next

def linked_list_to_list(head):
    """Convert a linked list to a Python list"""
    result = []
    
    while head:
        result.append(head.val)
        head = head.next
    
    return result

if __name__ == "__main__":
    solution = Solution()
    
    # Example 1
    print("Example 1:")
    l1 = list_to_linked_list([2, 4, 3])
    l2 = list_to_linked_list([5, 6, 4])
    print("Input: l1 = [2,4,3], l2 = [5,6,4]")
    result1 = solution.addTwoNumbers(l1, l2)
    output1 = linked_list_to_list(result1)
    print(f"Output: {output1}")
    expected1 = [7, 0, 8]
    print(f"Expected: {expected1}")
    print(f"Test passed: {output1 == expected1}")
    
    # Example 2
    print("\nExample 2:")
    l1 = list_to_linked_list([0])
    l2 = list_to_linked_list([0])
    print("Input: l1 = [0], l2 = [0]")
    result2 = solution.addTwoNumbers(l1, l2)
    output2 = linked_list_to_list(result2)
    print(f"Output: {output2}")
    expected2 = [0]
    print(f"Test passed: {output2 == expected2}")
    
    # Example 3
    print("\nExample 3:")
    l1 = list_to_linked_list([9, 9, 9, 9, 9, 9, 9])
    l2 = list_to_linked_list([9, 9, 9, 9])
    print("Input: l1 = [9,9,9,9,9,9,9], l2 = [9,9,9,9]")
    result3 = solution.addTwoNumbers(l1, l2)
    output3 = linked_list_to_list(result3)
    print(f"Output: {output3}")
    expected3 = [8, 9, 9, 9, 0, 0, 0, 1]
    print(f"Test passed: {output3 == expected3}")
    
    print("\nAll test cases completed!")